### PR TITLE
[IMP] lunch: scss revamp

### PR DESCRIPTION
**Pull request by morm-odoo**
Part of the overall v16 SCSS optimization/restyle, task-2704984.

task-2920029

### CODE REVIEW

File: addons/lunch/views/lunch_orders_views.xml [hunk 1]
```diff
@@ -216,57 +216,57 @@
...
                     <div class="flex-grow-1 pe-5">
                         <h2><field name="name"/></h2>
                         <h3 class="pt-3"><field name="price"/></h3>
                     </div>
                 </div>
                 <div class="o_lunch_wizard">
-                    <div class="row">
-                        <div class="col-2">
+                    <div class="grid gap-1">
+                        <div class="g-col-12 g-col-md-2">
```
Comment (stefanorigano) [inline addons/lunch/views/lunch_orders_views.xml:228]: Do not place labels on top, odoo's forms are always inline, in mobile too.

About the general layout, ideally this template should mirror the default `form-view` structure and classes, eg:
```xml

<table class="o_group o_inner_group">
    <tbody>
        <tr>
            <td class="o_td_label">
                <label class="o_form_label" for="o_field_input_60">Date</label>
            </td>
            <td style="width: 100%;">
                <span class="o_field_widget o_quick_editable" name="o_field_input_60">07/08/2022</span>
            </td>
        </tr>
    </tbody>
</table>
```
We'll probably switch to grid soon, I don't think it's worth to replicate the HTML structure, but we could use the same classes at least

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Thanks @morm-odoo , some remarks for you

File: addons/lunch/static/src/xml/lunch_templates.xml [hunk 2]
```diff
@@ -6,14 +6,14 @@
...
         </t>
-        <div class="o_lunch_banner container-fluid">
-            <div class="o_lunch_widget row py-3 py-md-0">
-                <div class="o_lunch_widget_info col-12 col-md-4 card border-0">
-                    <div class="card-body row g-0 align-items-center">
-                        <div class="col-3 col-md-6 col-lg-3">
+        <div class="o_lunch_banner container-fluid flex-shrink-0 border-bottom bg-white">
+            <div class="o_lunch_widget grid gap-1">
+                <div class="o_lunch_widget_info g-col-12 g-col-md-4">
+                    <div class="grid gap-1 flex-grow-1 align-items-center h-100 p-3 bg-white text-700">
```
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/lunch/static/src/xml/lunch_templates.xml:12]: `xxl` looks a bit off too, space is not optimised  
image

File: addons/web/static/src/legacy/scss/bootstrap_overridden.scss [hunk 3]
```diff
@@ -60,6 +60,9 @@ $variable-prefix: '' !default;
 // Restore negative margins disabled in BS5 by default
 $enable-negative-margins: true !default;
 
+// Enable CSS Grid classes
+$enable-cssgrid: true !default;
```
Comment (stefanorigano) [inline addons/web/static/src/legacy/scss/bootstrap_overridden.scss:64]: At first sight I don't see any advantage of using `grid` for this layout.
Should we reconsider it?

**Review by stefanorigano — CHANGES_REQUESTED** [review]

File: addons/lunch/static/src/scss/lunch_kanban.scss [hunk 4]
```diff
@@ -1,6 +1,6 @@
 .o_lunch_content {
     .o_legacy_kanban_view {
-        flex: 1 1 100%;
+        flex-basis: 100%;
```
Comment (stefanorigano) [inline addons/lunch/static/src/scss/lunch_kanban.scss:3]: we have `flex-basis-100`now :)

File: addons/lunch/static/src/js/lunch_controller_common.js [hunk 5]
```diff
@@ -42,7 +42,7 @@ var LunchControllerCommon = {
         // banner and renderer (this is required to get the desired
         // layout with the searchPanel to the left)
         var self = this;
-        this.$('.o_content').append($('<div>').addClass('o_lunch_content'));
+        this.$('.o_content').append($('<div>').addClass('o_lunch_content d-flex flex-column flex-grow-1 w-100 h-100'));
```
Comment (stefanorigano) [inline addons/lunch/static/src/js/lunch_controller_common.js:45]: Horizontal scrollbars should be in the table, not in the entire view (not sure if this is the line causing the issue)

image

File: addons/lunch/static/src/xml/lunch_templates.xml [hunk 6]
```diff
@@ -6,14 +6,14 @@
...
         </t>
-        <div class="o_lunch_banner container-fluid">
-            <div class="o_lunch_widget row py-3 py-md-0">
-                <div class="o_lunch_widget_info col-12 col-md-4 card border-0">
-                    <div class="card-body row g-0 align-items-center">
-                        <div class="col-3 col-md-6 col-lg-3">
+        <div class="o_lunch_banner container-fluid flex-shrink-0 border-bottom bg-white">
+            <div class="o_lunch_widget row">
+                <div class="o_lunch_widget_info col-12 col-lg-4">
+                    <div class="row flex-grow-1 align-items-center h-100 p-3 bg-white text-700">
```
Comment (stefanorigano) [inline addons/lunch/static/src/xml/lunch_templates.xml:12]: Are `bg-white text-700` needed?

File: addons/lunch/views/lunch_orders_views.xml [hunk 7]
```diff
@@ -216,57 +214,57 @@
                 <field name="available_today" invisible="1"/>
                 <div class="d-flex">
                     <div class="flex-grow-0 pe-5">
-                        <field name="image_1920" widget="image" class="o_lunch_image" options="{'image_preview': 'image_128'}"/>
+                        <field name="image_1920" widget="image" class="o_lunch_image"  options="{'image_preview': 'image_128'}"/>
```
Comment (stefanorigano) [inline addons/lunch/views/lunch_orders_views.xml:217]: typo/unnecessary change

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Ok, @morm-odoo , thanks.
There is a layout issue in list-view (check comment), for the rest just minors 👍

File: addons/lunch/static/src/scss/lunch_list.scss [hunk 8]
```diff
@@ -1,11 +1,7 @@
 .o_lunch_content {
-    .o_list_button {
-        width: 1px;
-    }
-
     .o_lunch_list_view {
         td:last-child {
             padding-right: 16px;
```
Comment (stefanorigano) [inline addons/lunch/static/src/scss/lunch_list.scss:4]: This should not be here :/
It's a workaround for an issue we have managing paddings in list-view.

Currently we have 2 design, one for standalone list-views and another one for embedded list-views (into form-view  or modals for example...).

You should fix it in form-view directly, [extending the selector](https://github.com/odoo/enterprise/blob/master/web_enterprise/static/src/legacy/scss/list_view.scss#L170) in order to match list view inside `o_controller_with_searchpanel`

**Review by stefanorigano — CHANGES_REQUESTED** [review]

**Comment by stefanorigano** [comment]
Ok, thanks @morm-odoo .

@adr-odoo , ready for tech review 👍 
cc @kba-odoo

**Review by stefanorigano — APPROVED** [review]

**Review by kbapt — APPROVED** [review]
LGTM :ok_hand: 

Thanks!

File: addons/lunch/static/src/xml/lunch_templates.xml [hunk 9]
```diff
@@ -28,30 +28,30 @@
...
                                     <t t-set="can_modify" t-value="!['sent', 'confirmed'].includes(line.raw_state)" />
                                     <div class="d-flex align-items-center">
-                                        <div class="flex-grow-0 flex-shrink-0 o_lunch_product_quantity">
-                                            <button class="btn btn-sm btn-icon btn-link fa fa-minus-circle o_remove_product" t-if="can_modify" t-attf-data-id="{{ line.id }}"/>
+                                        <div class="o_lunch_product_quantity flex-grow-0 flex-shrink-0">
+                                            <button class="o_remove_product btn btn-sm btn-icon btn-link p-0 fa fa-minus-circle" t-if="can_modify" t-attf-data-id="{{ line.id }}"/>
                                             <span t-esc="line.quantity"/>
-                                            <button class="btn btn-sm btn-icon btn-link fa fa-plus-circle o_add_product" t-if="can_modify" t-attf-data-id="{{ line.id }}"/>
+                                            <button class="o_add_product btn btn-sm btn-icon btn-link p-0 fa fa-plus-circle" t-if="can_modify" t-attf-data-id="{{ line.id }}"/>
```
Comment (pparidans) [inline addons/lunch/static/src/xml/lunch_templates.xml:46]: Could we take advantage of the Bootstrap `btn-group` component?

File: addons/lunch/static/src/xml/lunch_templates.xml [hunk 10]
```diff
@@ -75,17 +75,17 @@
                         </div>
                     </t>
                 </div>
-                <div class="o_lunch_widget_info col-12 col-md-4 card border-0">
+                <div class="o_lunch_widget_info col-12 col-lg-2">
                     <t t-if="!_.isEmpty(widget.lines) &amp;&amp; widget.raw_state == 'new'">
-                        <div class="card-body d-flex flex-column justify-content-between">
-                            <h4 class="card-title d-flex py-1">
+                        <div class="d-flex flex-column justify-content-between h-100 py-3 bg-white">
+                            <h4 class="d-flex py-1 mb-0 fw-bold">
```
Comment (pparidans) [inline addons/lunch/static/src/xml/lunch_templates.xml:81]: Isn't the `<h4>` already bold?

File: addons/lunch/static/src/xml/lunch_templates.xml [hunk 11]
```diff
@@ -6,14 +6,14 @@
...
+        <div class="o_lunch_banner container-fluid flex-shrink-0 border-bottom bg-white">
+            <div class="o_lunch_widget row">
+                <div class="o_lunch_widget_info col-12 col-lg-4">
+                    <div class="row flex-grow-1 align-items-center h-100 p-3">
+                        <div class="col-3">
                             <img class="o_image_64_cover rounded-circle" t-attf-src="{{ widget.userimage }}"/>
                         </div>
-                        <div class="col-9 col-md-6 col-lg-9">
+                        <div class="col-9 col-md-6 col-lg-9 flex-grow-1">
                             <div class="ps-3">
```
Comment (pparidans) [inline addons/lunch/static/src/xml/lunch_templates.xml:17]: Could we merge those two node into one?

**Review by pparidans — CHANGES_REQUESTED** [review]
A general remark:

could you check if all the `flex-grow-*` and `flex-shrink-*` are really needed because, for a certain number of them, I have a hard time seeing the difference with or without them... (and of course, they are useless when the parent isn't `flex` :wink: ).

File: addons/lunch/static/src/scss/lunch_view.scss [hunk 12]
```diff
@@ -1,77 +1,3 @@
...
-}
-
-.o_lunch_wizard {
-    .col-10 {
-        .o_form_label {
-            font-weight: normal !important;
-        }
-    }
+.o_lunch_widget_lines{
+    max-height: (($btn-font-size * $btn-line-height) + 3) * 2.5;
```
[the remark this answers is not in the document]
Comment (morm-odoo) [reply addons/lunch/static/src/scss/lunch_view.scss:2]: max-height: ((font-size * line-height) + (border size top + border size bottom + padding from div wrapping)) * (the number of lines to show/2);
Here it shows 5 lines/orders and then scroll.
I added the comment explaining it back.

**Comment by morm-odoo** [comment]
@stefanorigano

File: addons/lunch/static/src/scss/lunch_view.scss [hunk 13]
```diff
@@ -1,58 +1,8 @@
...
-    }
-}
-
-.o_lunch_image {
-    img {
-        max-width: 128px;
-        max-height: 128px !important;
-    }
+.o_lunch_widget_lines{
+    //calcul to know the number of items to show before scrolling
```
Comment (stefanorigano) [inline addons/lunch/static/src/scss/lunch_view.scss:2]: typo ;)

Comment (stefanorigano) [inline addons/lunch/static/src/scss/lunch_view.scss:2]: typo ;)

Comment (stefanorigano) [inline addons/lunch/static/src/scss/lunch_view.scss:2]: typo ;)

File: addons/lunch/static/src/scss/lunch_view.scss (hunk 12 above)
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/lunch/static/src/scss/lunch_view.scss:4]: Since the tallest element of an entry is the title (and not the button), I'd say;
```scss
$-entry-height: calc(#{$font-size-base * $line-height-base} + (#{map-get($spacers, 1)} * 2));

max-height: calc(#{$-entry-height} * 5); // if we want 5 entries
```
I'd suggest 5,5 actually. Doing so  the beginning of the 6th row will be visible and the user understands that he has to scroll.

File: addons/lunch/static/src/scss/lunch_view.scss [hunk 14]
```diff
@@ -1,77 +1,21 @@
...
-}
-
-.o_lunch_image {
-    img {
-        max-width: 128px;
-        max-height: 128px !important;
     }
 }
 
 @include media-breakpoint-down(md) {
```
[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/lunch/static/src/scss/lunch_view.scss:6]: have you check?

**Review by stefanorigano — CHANGES_REQUESTED** [review]
@morm-odoo , some remarks ;)

File: addons/lunch/static/src/scss/lunch_view.scss (hunk 14 above)
[the remark this answers is not in the document]
Comment (morm-odoo) [reply addons/lunch/static/src/scss/lunch_view.scss:6]: It's linked to the caret on iOS, so I guess it should stay but what do you mean by the template is loaded in js for mobile only?

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/lunch/static/src/scss/lunch_view.scss:6]: ok, let's keep it

**Comment by pparidans** [comment]
An additional remark, when a product has options, the option's price is misaligned with its product's price.
image

**Comment by pparidans** [comment]
About the order's line `max-height`, limiting the height is definitely a good idea :+1:  but, from a design stand point, wouldn't it be better to not have the banner's height change depending on the content of the user's order? Not sure as it would need to be bigger than currently (aka. wasted space for small orders) but more predictable when adding more products... :thinking:

**Comment by morm-odoo** [comment]
@pparidans 
Screenshot 2022-08-29 at 11 26 14
Screenshot 2022-08-29 at 11 24 29

When we add a lot of products it's not great (it was the previous behaviour), with a double scroll here. 
Also,
- If the header is allowed to get as big as the content it won't be consistent
- There can be massive spaces above/below the user in the left column
- There can be a massive space between the total and the order button
- Content of the whole page will continuously get pushed down as the header gets higher

**Comment by stefanorigano** [comment]
> About the order's line `max-height`, limiting the height is definitely a good idea 👍 but, from a design stand point, wouldn't it be better to not have the banner's height change depending on the content of the user's order? Not sure as it would need to be bigger than currently (aka. wasted space for small orders) but more predictable when adding more products... 🤔

@pparidans , I think that  for small orders (2 items) the height's difference is barely noticeable and it's kind-of needed.
If we trigger the scrollbar to fit the default header size people may miss that a 3rd item has been added. 
Honestly I have some remarks about this design, it could be improved... but it's a bit out of scope for this tasks.

@morm-odoo  don't miss https://github.com/odoo/odoo/pull/96027#issuecomment-1229988909 ;)

**Comment by pparidans** [comment]
@stefanorigano @morm-odoo To be clear, I'm not defending the previous "let's grow as much as needed even if the whole layout is broken" behavior. Limiting the height is great. I was only wondering if a fixed height wouldn't be better (cf. a more "stable" design ; no grow, no shrink)?

But in the end, I leave you both the final word on this because it's definitively better than before :hugs:

**Comment by pparidans** [comment]
@morm-odoo Also, don't forget the tests :wink:

**Comment by pparidans** [comment]
The error in the enterprise tests looked a bit weird... so I rebuilded it with automatic-rebase. We'll see...

On another topic, the cart on mobile has some alignment issues:
18814429-master-all runbot60 odoo com_web(Runbot Mobile)

File: addons/lunch/views/lunch_orders_views.xml [hunk 15]
```diff
@@ -224,49 +222,49 @@
...
-                        <div class="col-2">
-                            <field name="topping_label_1" nolabel="1" attrs="{'invisible': [('available_toppings_1', '=', False)]}" class="fw-bold"/>
+                    <div class="row py-3 py-md-0">
+                        <div class="o_td_label col-3 col-md-2">
+                            <field name="topping_label_1" nolabel="1" attrs="{'invisible': [('available_toppings_1', '=', False)]}" class="o_form_label"/>
                         </div>
-                        <div class="col-10">
-                            <field name="topping_ids_1" attrs="{'invisible': [('available_toppings_1', '=', False)]}" widget="many2many_checkboxes" nolabel="1" domain="[('topping_category', '=', 1), ('supplier_id', '=', supplier_id)]"/>
+                        <div class="col-9 col-md-10">
+                            <field name="topping_ids_1" attrs="{'invisible': [('available_toppings_1', '=', False)]}" widget="many2many_checkboxes" nolabel="1" domain="[('topping_category', '=', 1), ('supplier_id', '=', supplier_id)]" class="o_field_widget o_quick_editable" />
```
Comment (pparidans) [inline addons/lunch/views/lunch_orders_views.xml:230]: It tastes like `<group>`, it smells like `<group>` but it's not an actual `<group>`... and I have to admit, I really don't like that. But in this case (cf. having a field as the label for another field), there is no other (that I know of) way to do it :sob: 
with the current tools provided by the framework. Probably something to discuss with them at some point, but out of this task's scope so, let's keep it that way for now :shrug: ...

**Comment by pparidans** [comment]
I'm sorry @morm-odoo , but you will need to rebase your branch as https://github.com/odoo/odoo/pull/99021 has just been merged... :cry:

**Comment by pparidans** [comment]
LGTM :sunglasses: 
@robodoo r+
